### PR TITLE
Cloudkitty role config

### PR DESCRIPTION
**Pull request by MaxBed4d**
Added some config in order to enable compatibility with the new Cloudkitty Ansible role, which enables Cloudkitty to be configured as part of the `openstack-config` deployment.

### CODE REVIEW

File: requirements.yml
Comment (MaxBed4d) [inline requirements.yml:1]: This needs to be reviewed once the Cloudkitty Ansible role PR has been approved and merged.

File: examples/ratings.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+# Example configuration for registering Cloudkitty fields.
+# Included is mapping for various different compute flavors & 
+# a service mapping based on the stored image size in Glance. 
+
+ratings_hashmap_field_mappings:
+  - service: instance
+    name: flavor_id
+    mappings:
+      - value: '1' #tiny
```
Comment (MoteHue) [inline examples/ratings.yml:13]: It would be good to mention that these values are flavour ids.

File: ansible/group_vars/all/openstack [hunk 2]
```diff
@@ -48,6 +48,28 @@ openstack_image_elements: []
...
+###############################################################################
+# Configuration variables for a CloudKitty ratings service deployment.
+
+# File path or URL containing Python upper constraints.
+#
+ratings_upper_constraints_file: []
+
+# Environment variables for the OpenStack CLI. By default this is left empty.
+#
+ratings_environment: []
```
Comment (MoteHue) [inline ansible/group_vars/all/openstack:60]: I don't think you need to define these here, the defaults from the role should be sufficient.

File: ansible/group_vars/all/openstack [hunk 3]
```diff
@@ -48,6 +48,28 @@ openstack_image_elements: []
...
+# with the 'mappings' field also being a list of dictionaries. 
+# Example of the mappings and their fields can be found below, 
+# however for more information please refer to the README.md file.
+#
+ratings_hashmap_field_mappings: []
+
+# Much like the field mappings above, the service mappings are a list of
+# dictionaries, however these are not associated with a field.
+#
+ratings_hashmap_service_mappings: []
```
Comment (MoteHue) [inline ansible/group_vars/all/openstack:72]: Can you rename these with the prefix ``openstack_`` so they match the other roles?

File: ansible/openstack-ratings.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+- name: Generate OpenStack software ratings
+  hosts: localhost
+  tags:
+    - ratings
+  roles:
+    - role: stackhpc.openstack.os_ratings
+      os_ratings_hashmap_field_mappings: "{{ ratings_hashmap_field_mappings }}"
+      os_ratings_hashmap_service_mappings: "{{ ratings_hashmap_service_mappings }}"
+      os_ratings_venv: "{{ openstack_venv }}"
+      os_ratings_upper_constraints_file: "{{ ratings_upper_constraints_file }}"
```
Comment (MoteHue) [inline ansible/openstack-ratings.yml:11]: Can drop overriding ``os_ratings_upper_constraints_file``, the default is sufficient.

File: etc/openstack-config/openstack-config.yml [hunk 5]
```diff
@@ -96,6 +96,48 @@
...
+
+# File path or URL containing Python upper constraints.
+####### THIS IS NOT NECCESSARY TO CONFIGURE ###########
+#ratings_upper_constraints_file: /path/to/upper_constraints_file
+
+# Environment variables for the OpenStack CLI. By default this is left empty.
+################### THIS IS NOT NECCESSARY TO CONFIGURE #####################
+#ratings_environment:
+#  - name: ENV_VAR_NAME
+#    value: ENV_VAR_VALUE
```
Comment (MoteHue) [inline etc/openstack-config/openstack-config.yml:116]: Likewise I think you can drop this here.

**Review by MoteHue — CHANGES_REQUESTED** [review]

File: ansible/group_vars/all/openstack (hunk 2 above)
Comment (MaxBed4d) [reply ansible/group_vars/all/openstack:60]: These will error out as the `openstack-ratings.yml` playbook will request these variables before the role is called upon and being able to provide values. Additionally, giving the user the option to include them means that they should also provide a default.

File: ansible/group_vars/all/openstack (hunk 3 above)
Comment (MaxBed4d) [reply ansible/group_vars/all/openstack:72]: By default, these are the actual variable names as used by CloudKitty & OpenStack, and this is the convention used by the other services.

File: ansible/openstack-ratings.yml (hunk 4 above)
Comment (MaxBed4d) [reply ansible/openstack-ratings.yml:11]: I still think providing the option is better than locking it. Especially as this is going to be where CloudKitty is almost solely going to be configured.

Comment (MoteHue) [reply ansible/openstack-ratings.yml]: I was thinking along the lines of matching the conventions the other roles use. I can't immediately see why we'd want to override these values often, but I guess that it doesn't hurt to expose them.

File: ansible/group_vars/all/openstack (hunk 2 above)
Comment (MoteHue) [reply ansible/group_vars/all/openstack]: Responded in the other comment thread:
> I was thinking along the lines of matching the conventions the other roles use. I can't immediately see why we'd want to override these values often, but I guess that it doesn't hurt to expose them.

**Review by MoteHue — APPROVED** [review]

File: ansible/openstack-ratings.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  hosts: localhost
+  tags:
+    - ratings
+  roles:
+    - role: stackhpc.openstack.os_ratings
+      os_ratings_hashmap_field_mappings: "{{ openstack_ratings_hashmap_field_mappings }}"
+      os_ratings_hashmap_service_mappings: "{{ openstack_ratings_hashmap_service_mappings }}"
+      os_ratings_venv: "{{ openstack_venv }}"
+      os_ratings_upper_constraints_file: "{{ openstack_ratings_upper_constraints_file }}"
+      os_ratings_environment: "{{ openstack_ratings_environment }}"
```
Comment (markgoddard) [inline ansible/openstack-ratings.yml:12]: I don't think these two are necessary. The default upper constraints file should be fine. It's possible we'd need to set the environment in specific cases, but I think we can leave it unset for now. Or if we do want to set it then it should be to a shared `openstack_environment` variable with a default of `{}` in `ansible/group_vars/all/openstack-auth`.

File: etc/openstack-config/openstack-config.yml [hunk 7]
```diff
@@ -96,6 +96,48 @@
...
+# File path or URL containing Python upper constraints.
+####### THIS IS NOT NECCESSARY TO CONFIGURE ###########
+#openstack_ratings_upper_constraints_file: /path/to/upper_constraints_file
+
+# Environment variables for the OpenStack CLI. By default this is left empty.
+################### THIS IS NOT NECCESSARY TO CONFIGURE #####################
+#openstack_ratings_environment:
+#  - name: ENV_VAR_NAME
+#    value: ENV_VAR_VALUE
+
```
Comment (markgoddard) [inline etc/openstack-config/openstack-config.yml:117]: Not required

File: etc/openstack-config/openstack-config.yml [hunk 8]
```diff
@@ -96,6 +96,48 @@
...
+# A list where each item is a dictionary mapping the associated fields, 
+# with the 'mappings' field also being a list of dictionaries. 
+# Example of the mappings and their fields can be found below, 
+# however for more information please refer to the README.md file.
```
Comment (markgoddard) [inline etc/openstack-config/openstack-config.yml:121]: ... in the stackhpc.openstack.os_ratings role

File: etc/openstack-config/openstack-config.yml [hunk 9]
```diff
@@ -96,6 +96,48 @@
...
+#
+#openstack_ratings_hashmap_field_mappings:
+#  - service: SERVICE_NAME
+#    name: FIELD_NAME
+#    mappings:
+#      - value: MAPPING_VALUE
+#        cost: MAPPING_COST
+#        group: MAPPING_GROUP
```
Comment (markgoddard) [inline etc/openstack-config/openstack-config.yml:129]: Mark as optional?

File: etc/openstack-config/openstack-config.yml [hunk 10]
```diff
@@ -96,6 +96,48 @@
...
+#        type: MAPPING_TYPE
+
+# Much like the field mappings above, the service mappings are a list of
+# dictionaries, however these are not associated with a field.
+#
+#openstack_ratings_hashmap_service_mappings:
+#  - service: SERVICE_NAME
+#    cost: MAPPING_COST
+#    group: MAPPING_GROUP
```
Comment (markgoddard) [inline etc/openstack-config/openstack-config.yml:138]: Mark as optional?

File: examples/ratings.yml [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+# #   #   #   #     #     #        'cloudkitty.conf' & 'metrics.yml' ARE    
+#  #  #   #   #     #     ###      CONFIGURED AS SHOWN BELOW. MORE INFO CAN
+#   # #   #   #     #     #        BE FOUND IN THE 'stackhpc-kayobe-config' 
+#     #    ###      #     #####    DOCS.
+
+
+# CORRESPONDING cloudkitty.conf
+#
+# [collect]
+# period = 120
```
Comment (markgoddard) [inline examples/ratings.yml:54]: Not really relevant

File: examples/ratings.yml [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+# scope_key = tenant_id
+
+# CORRESPONDING metrics.yml
+#
+# metrics:
```
Comment (markgoddard) [inline examples/ratings.yml:59]: I expect these metrics are specific to the prometheus collector with the openstack exporter.

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: ansible/group_vars/all/openstack [hunk 13]
```diff
@@ -48,6 +48,28 @@ openstack_image_elements: []
 
 # List of diskimage-builder element git repos to pull
 openstack_image_git_elements: []
+###############################################################################
```
[the remark this answers is not in the document]
Comment (MaxBed4d) [reply ansible/group_vars/all/openstack:51]: Moved and resolved.

File: ansible/openstack-ratings.yml (hunk 6 above)
Comment (MaxBed4d) [reply ansible/openstack-ratings.yml:12]: Removed them.

File: etc/openstack-config/openstack-config.yml (hunk 8 above)
Comment (MaxBed4d) [reply etc/openstack-config/openstack-config.yml:121]: Amended and edited.

File: examples/ratings.yml (hunk 12 above)
Comment (MaxBed4d) [reply examples/ratings.yml:59]: Removed.

File: examples/ratings.yml (hunk 11 above)
Comment (MaxBed4d) [reply examples/ratings.yml:54]: Removed.

File: etc/openstack-config/openstack-config.yml (hunk 9 above)
Comment (MaxBed4d) [reply etc/openstack-config/openstack-config.yml:121]: I thought it was only the `group` which was optional, however GitHub showed the comment as highlighting and reffering to the entire `mapping` so I thought I simply misunderstood.

File: ansible/group_vars/all/openstack [hunk 14]
```diff
@@ -62,3 +62,26 @@ openstack_host_aggregates: []
...
+# Configuration variables for a CloudKitty ratings service deployment.
+
+# File path or URL containing Python upper constraints.
+#
+openstack_ratings_upper_constraints_file: []
+
+# Environment variables for the OpenStack CLI. By default this is left empty.
+#
+openstack_ratings_environment: []
+
```
[the remark this answers is not in the document]
Comment (MaxBed4d) [reply ansible/group_vars/all/openstack:76]: True, forgot about that one!

**Review by markgoddard — APPROVED** [review]
Thanks!

**Pull request merged**